quinn/src/connection.rs

```diff
@@ -660,49 +657,30 @@ impl ConnectionInner {
         }
     }
 
-    fn drive_timers(&mut self, cx: &mut Context, now: Instant) -> bool {
+    fn drive_timer(&mut self, cx: &mut Context, now: Instant) -> bool {
         let mut keep_going = false;
-        for (timer, slot) in &mut self.timers {
-            if let Some(ref mut delay) = slot {
-                match delay.poll_unpin(cx) {
-                    Poll::Ready(()) => {
-                        *slot = None;
-                        trace!("{:?} timeout", timer);
-                        self.inner.handle_timeout(now, timer);
-                        // Timeout call may have queued sends
-                        keep_going = true;
-                    }
-                    Poll::Pending => {}
+        if self.timer.is_none() {
+            self.timer = self
+                .inner
+                .poll_timeout()
+                .map(|x| delay_until(TokioInstant::from_std(x)));
+        }
+        while let Some(ref mut delay) = self.timer {
+            match delay.poll_unpin(cx) {
```

Comment:
I think `if let` could be nicer here for the early return -- since we don't need the value from `Ready`, we can dedent the `Ready` block contents by two levels.